File: cubecli/commands/project.py
import typer
from typing import Optional
from cubecli.api_client import APIClient
from cubecli.utils import (
    print_success, print_error, print_json, create_table,
    confirm_action, get_context_value, with_spinner, format_status, console,
    handle_api_exception
)
import httpx

app = typer.Typer(no_args_is_help=True)

@app.command()
def create(
    ctx: typer.Context,
    name: str = typer.Option(..., "--name", "-n", help="Project name"),
    description: Optional[str] = typer.Option(None, "--description", "-d", help="Project description"),
):
    """Create a new project"""
    api_token = get_context_value(ctx, "api_token")
    json_output = get_context_value(ctx, "json", False)
    
    if not api_token:
        print_error("No API token configured")
        raise typer.Exit(1)
    
    client = APIClient(api_token)
    
    data = {"name": name}
    if description:
        data["description"] = description
    
    with with_spinner("Creating project...") as progress:
        task = progress.add_task("Creating project...", total=None)
        try:
            response = client.post("/projects/", data)
            progress.update(task, completed=True)
        except Exception as e:
            handle_api_exception(e, progress)
    
    if json_output:
        print_json(response)
    else:
        print_success(f"Project '{name}' created successfully!")
        print_success(f"Project ID: {response.get('id', 'N/A')}")

@app.command("list")
def list_projects(ctx: typer.Context):
    """List all projects"""
    api_token = get_context_value(ctx, "api_token")
    json_output = get_context_value(ctx, "json", False)
    
    if not api_token:
        print_error("No API token configured")
        raise typer.Exit(1)
    
    client = APIClient(api_token)
    
    with with_spinner("Fetching projects...") as progress:
        task = progress.add_task("Fetching projects...", total=None)
        try:
            response = client.get("/projects/")
            progress.update(task, completed=True)
        except Exception as e:
            handle_api_exception(e, progress)
    
    # The API returns a list directly, not a dict with "projects" key
    projects = response if isinstance(response, list) else response.get("projects", [])
    
    if json_output:
        print_json(projects)
    else:
        if not projects:
            print_error("No projects found")
            return
        
        table = create_table("Projects", ["ID", "Name", "Description", "VPS Count", "Networks", "Created"])
        
        for item in projects:
            # Extract project info from the structure
            project = item.get("project", {})
            vps_count = len(item.get("vps", []))
            networks_count = len(item.get("networks", []))
            
            description = project.get("description", "N/A")
            if description and len(description) > 30:
                description = description[:30] + "..."
            
            table.add_row(
                str(project.get("id", "N/A")),
                project.get("name", "N/A"),
                description,
                str(vps_count),
                str(networks_count),
                project.get("created_at", "N/A")
            )
        
        console.print()
        console.print(table)

@app.command()
def show(
    ctx: typer.Context,
    project_id: int = typer.Argument(..., help="Project ID to show"),
):
    """Show detailed project information"""
    api_token = get_context_value(ctx, "api_token")
    json_output = get_context_value(ctx, "json", False)
    
    if not api_token:
        print_error("No API token configured")
        raise typer.Exit(1)
    
    client = APIClient(api_token)
    
    with with_spinner("Fetching project details...") as progress:
        task = progress.add_task("Fetching project details...", total=None)
        try:
            response = client.get("/projects/")
            progress.update(task, completed=True)
        except Exception as e:
            handle_api_exception(e, progress)
    
    # The API returns a list directly
    projects = response if isinstance(response, list) else response.get("projects", [])
    
    # Find the specific project
    project_data = None
    for item in projects:
        proj = item.get("project", {})
        if proj.get("id") == project_id:
            project_data = item
            break
    
    if not project_data:
        print_error(f"Project {project_id} not found")
        raise typer.Exit(1)
    
    project = project_data.get("project", {})
    
    if json_output:
        print_json(project_data)
    else:
        console.print()
        console.print(f"[bold]Project: {project.get('name', 'N/A')}[/bold]")
        console.print(f"ID: {project.get('id', 'N/A')}")
        console.print(f"Description: {project.get('description', 'N/A')}")
        console.print(f"Created: {project.get('created_at', 'N/A')}")
        
        # Show VPS if any
        vps_list = project_data.get("vps", [])
        if vps_list:
            console.print()
            vps_table = create_table(f"VPS in Project ({len(vps_list)})", ["ID", "Name", "Status", "IP", "Plan"])
            for vps in vps_list:
                # Extract floating IP
                floating_ip = vps.get("floating_ip", {})
                ip_address = floating_ip.get("address", "N/A")
                
                # Extract plan info
                plan = vps.get("plan", {})
                plan_name = plan.get("plan_name", "N/A")
                
                vps_table.add_row(
                    str(vps.get("id", "N/A")),
                    vps.get("name", "N/A"),
                    format_status(vps.get("status", "unknown")),
                    ip_address,
                    plan_name
                )
            console.print(vps_table)
        
        # Show networks if any
        networks = project_data.get("networks", [])
        if networks:
            console.print()
            net_table = create_table(f"Networks in Project ({len(networks)})", ["ID", "Name", "IP Range", "Location"])
            for net in networks:
                net_table.add_row(
                    str(net.get("id", "N/A")),
                    net.get("name", "N/A"),
                    f"{net.get('ip_range', 'N/A')}/{net.get('prefix', 'N/A')}",
                    net.get("location_name", "N/A")
                )
            console.print(net_table)

@app.command()
def delete(
    ctx: typer.Context,
    project_id: int = typer.Argument(..., help="Project ID to delete"),
    force: bool = typer.Option(False, "--force", "-f", help="Skip confirmation"),
):
    """Delete a project"""
    api_token = get_context_value(ctx, "api_token")
    json_output = get_context_value(ctx, "json", False)
    
    if not api_token:
        print_error("No API token configured")
        raise typer.Exit(1)
    
    if not force:
        if not confirm_action(f"Are you sure you want to delete project {project_id}?"):
            print_error("Operation cancelled")
            return
    
    client = APIClient(api_token)
    
    with with_spinner("Deleting project...") as progress:
        task = progress.add_task("Deleting project...", total=None)
        try:
            response = client.delete(f"/projects/{project_id}")
            progress.update(task, completed=True)
        except Exception as e:
            handle_api_exception(e, progress)
    
    if json_output:
        print_json({"message": "Project deleted successfully", "project_id": project_id})
    else:
        print_success(f"Project {project_id} deleted successfully!")

File: cmd/project/project.go
package project

import (
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/CubePathInc/cubecli/internal/cmdutil"
	"github.com/CubePathInc/cubecli/internal/output"
	"github.com/spf13/cobra"
)

func NewCmd() *cobra.Command {
	projectCmd := &cobra.Command{
		Use:   "project",
		Short: "Manage projects",
	}

	projectCreateCmd := &cobra.Command{
		Use:   "create",
		Short: "Create a new project",
		RunE: func(cmd *cobra.Command, args []string) error {
			client := cmdutil.GetClient(cmd)

			name, _ := cmd.Flags().GetString("name")
			description, _ := cmd.Flags().GetString("description")

			body := map[string]string{"name": name}
			if description != "" {
				body["description"] = description
			}

			s := output.NewSpinner("Creating project...")
			s.Start()
			resp, err := client.Post("/projects/", body)
			s.Stop()
			if err != nil {
				return err
			}

			if cmdutil.IsJSON(cmd) {
				return output.PrintJSON(json.RawMessage(resp))
			}

			output.PrintSuccess("Project created successfully")
			return nil
		},
	}

	projectListCmd := &cobra.Command{
		Use:   "list",
		Short: "List projects",
		RunE: func(cmd *cobra.Command, args []string) error {
			client := cmdutil.GetClient(cmd)

			s := output.NewSpinner("Fetching projects...")
			s.Start()
			resp, err := client.Get("/projects/")
			s.Stop()
			if err != nil {
				return err
			}

			if cmdutil.IsJSON(cmd) {
				return output.PrintJSON(json.RawMessage(resp))
			}

			var projects []struct {
				Project struct {
					ID          int    `json:"id"`
					Name        string `json:"name"`
					Description string `json:"description"`
				} `json:"project"`
				VPS      []json.RawMessage `json:"vps"`
				Networks []json.RawMessage `json:"networks"`
			}
			if err := json.Unmarshal(resp, &projects); err != nil {
				return fmt.Errorf("failed to parse response: %w", err)
			}

			t := output.NewTable("Projects", []string{"ID", "Name", "Description", "VPS Count", "Networks"})
			for _, p := range projects {
				t.AddRow(
					strconv.Itoa(p.Project.ID),
					p.Project.Name,
					p.Project.Description,
					strconv.Itoa(len(p.VPS)),
					strconv.Itoa(len(p.Networks)),
				)
			}
			t.Render()
			return nil
		},
	}

	projectShowCmd := &cobra.Command{
		Use:   "show <project_id>",
		Short: "Show project details",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			client := cmdutil.GetClient(cmd)

			projectID, err := strconv.Atoi(args[0])
			if err != nil {
				return fmt.Errorf("invalid project_id: %s", args[0])
			}

			s := output.NewSpinner("Fetching project details...")
			s.Start()
			resp, err := client.Get("/projects/")
			s.Stop()
			if err != nil {
				return err
			}

			var projects []struct {
				Project struct {
					ID          int    `json:"id"`
					Name        string `json:"name"`
					Description string `json:"description"`
				} `json:"project"`
				VPS []struct {
					ID       int    `json:"id"`
					Hostname string `json:"hostname"`
					Status   string `json:"status"`
				} `json:"vps"`
				Networks []struct {
					ID       int    `json:"id"`
					Name     string `json:"name"`
					IPRange  string `json:"ip_range"`
					Location string `json:"location_name"`
				} `json:"networks"`
			}
			if err := json.Unmarshal(resp, &projects); err != nil {
				return fmt.Errorf("failed to parse response: %w", err)
			}

			var found bool
			for _, p := range projects {
				if p.Project.ID == projectID {
					found = true

					if cmdutil.IsJSON(cmd) {
						return output.PrintJSON(p)
					}

					t := output.NewTable("Project Details", []string{"Field", "Value"})
					t.AddRow("ID", strconv.Itoa(p.Project.ID))
					t.AddRow("Name", p.Project.Name)
					t.AddRow("Description", p.Project.Description)
					t.AddRow("VPS Count", strconv.Itoa(len(p.VPS)))
					t.AddRow("Networks", strconv.Itoa(len(p.Networks)))
					t.Render()

					if len(p.VPS) > 0 {
						vt := output.NewTable("VPS Instances", []string{"ID", "Hostname", "Status"})
						for _, v := range p.VPS {
							vt.AddRow(strconv.Itoa(v.ID), v.Hostname, output.FormatStatus(v.Status))
						}
						vt.Render()
					}

					if len(p.Networks) > 0 {
						nt := output.NewTable("Networks", []string{"ID", "Name", "IP Range", "Location"})
						for _, n := range p.Networks {
							nt.AddRow(strconv.Itoa(n.ID), n.Name, n.IPRange, n.Location)
						}
						nt.Render()
					}

					break
				}
			}

			if !found {
				return fmt.Errorf("project with ID %d not found", projectID)
			}

			return nil
		},
	}

	projectDeleteCmd := &cobra.Command{
		Use:   "delete <project_id>",
		Short: "Delete a project",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			client := cmdutil.GetClient(cmd)

			projectID, err := strconv.Atoi(args[0])
			if err != nil {
				return fmt.Errorf("invalid project_id: %s", args[0])
			}

			if !cmdutil.CheckForce(cmd, "Are you sure you want to delete this project?") {
				output.PrintWarning("Aborted")
				return nil
			}

			s := output.NewSpinner("Deleting project...")
			s.Start()
			resp, err := client.Delete(fmt.Sprintf("/projects/%d", projectID))
			s.Stop()
			if err != nil {
				return err
			}

			if cmdutil.IsJSON(cmd) {
				return output.PrintJSON(json.RawMessage(resp))
			}

			output.PrintSuccess("Project deleted successfully")
			return nil
		},
	}

	projectCreateCmd.Flags().StringP("name", "n", "", "Name of the project")
	projectCreateCmd.Flags().StringP("description", "d", "", "Description of the project")
	projectCreateCmd.MarkFlagRequired("name")

	projectDeleteCmd.Flags().BoolP("force", "f", false, "Skip confirmation prompt")

	projectCmd.AddCommand(projectCreateCmd, projectListCmd, projectShowCmd, projectDeleteCmd)
	return projectCmd
}
